go-controller/pkg/clustermanager/endpointslicemirror/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,428 @@
+			_, err := c.kubeClient.DiscoveryV1().EndpointSlices(namespace).Create(ctx, currentMirror, metav1.CreateOptions{})
+			return err
+		}
+		klog.Infof("Updating the mirrored EndpointSlice: %s for: %s", cache.MetaObjectToName(currentMirror), cache.MetaObjectToName(defaultEndpointSlice))
+		_, err := c.kubeClient.DiscoveryV1().EndpointSlices(namespace).Update(ctx, currentMirror, metav1.UpdateOptions{})
+		return err
+	}
+	return nil
+}
+
+// isManagedByController determines if the provided endpointSlice is managed by the current controller by checking the
+// "endpointslice.kubernetes.io/managed-by" label value.
+func (c *Controller) isManagedByController(endpointSlice *v1.EndpointSlice) bool {
+	return c.name == endpointSlice.Labels[v1.LabelManagedBy]
+}
+
+// isManagedByController determines if the provided endpointSlice is managed by the default EndpointSlice
+// controller by checking the "endpointslice.kubernetes.io/managed-by" label value.
+func isManagedByDefault(endpointSlice *v1.EndpointSlice) bool {
+	return types.EndpointSliceDefaultControllerName == endpointSlice.Labels[v1.LabelManagedBy]
+}
+
+// getPodIP retrieves the IP address of a specified Pod within a given namespace and network.
+// If the pod is host networked it returns default pod IP from the status ignoring the network.
+// Otherwise, it unmarshals the Pod's network annotation, and matches the IP from the provided network.
+func (c *Controller) getPodIP(name, namespace, network string, isIPv6 bool) (string, error) {
+	var podIP string
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		return "", err
+	}
+
+	if pod.Spec.HostNetwork {
+		podIPs, err := util.DefaultNetworkPodIPs(pod)
+		if err != nil {
+			return "", err
+		}
+
+		// MatchFirstIPFamily returns an error if no IP was found
+		ipAddr, err := util.MatchFirstIPFamily(isIPv6, podIPs)
+		if err != nil {
+			return "", err
+		}
+
+		podIP = ipAddr.String()
+	} else {
+		net, err := util.UnmarshalPodAnnotation(pod.Annotations, network)
+		if err != nil {
+			return "", err
+		}
+
+		// MatchFirstIPFamily returns an error if no IP was found
+		ipNet, err := util.MatchFirstIPNetFamily(isIPv6, net.IPs)
+		if err != nil {
+			return "", err
+		}
+		podIP = ipNet.IP.String()
+	}
+
+	return podIP, nil
+}
+
+// mirroredEndpointSlice creates or updates a mirrored EndpointSlice based on the provided defaultEndpointSlice.
+// The mirrored EndpointSlice will have custom labels set and will be managed by the current controller.
+func (c *Controller) mirroredEndpointSlice(mirroredEndpointSlice, defaultEndpointSlice *v1.EndpointSlice, network string) *v1.EndpointSlice {
+	var currentMirror *v1.EndpointSlice
+	if mirroredEndpointSlice != nil {
+		currentMirror = mirroredEndpointSlice.DeepCopy()
+	}
+	if currentMirror == nil {
+		currentMirror = &v1.EndpointSlice{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:       defaultEndpointSlice.Namespace,
+				OwnerReferences: defaultEndpointSlice.OwnerReferences,
```

Comment:
Yes, it would get removed anyway as a reaction to the default EndpointSlice begin removed but cloning the ownerReferece should help in potential corner cases where our controller is not running and the service is removed.